src/tir/transforms/inject_permuted_layout.cc

```diff
@@ -77,183 +89,196 @@ class PermutedLayoutInjector : public StmtExprMutator {
+    if (scope.rank != StorageRank::kShared) {
+      return store;
+    }
+
+    auto store_node = store.CopyOnWrite();
+    store_node->indices = HandleBufferIndices(store_node->buffer, store_node->indices);
+    return store;
+  }
+
+  PrimExpr VisitExpr_(const BufferLoadNode* op) final {
+    // Rewrite load from shared or shared.dyn to global
+    auto load = Downcast<BufferLoad>(IRMutatorWithAnalyzer::VisitExpr_(op));
+
+    if (!permute_ || load->buffer->shape.size() < 2) {
+      return load;
+    }
+
+    auto scope = StorageScope::Create(GetPtrStorageScope(load->buffer->data));
+    if (scope.rank != StorageRank::kShared) {
+      return load;
+    }
+
+    auto load_node = load.CopyOnWrite();
+    load_node->indices = HandleBufferIndices(load_node->buffer, load_node->indices);
+    return load;
+  }
+
+  PrimExpr HandleAccessPtrAndOffset(PrimExpr access_ptr, Optional<PrimExpr> offset = NullOpt) {
+    // The 2th arg of T.tvm_access_ptr call is offset, we set it to 0 and accumulate it to
+    // smem_offset
+    CHECK(access_ptr->IsInstance<CallNode>())
+        << "Invalid access ptr for permuted layout: " << access_ptr;
+    auto access_ptr_call = Downcast<Call>(access_ptr);
+    CHECK(access_ptr_call->op.same_as(builtin::tvm_access_ptr()))
+        << "Invalid access ptr for permuted layout: " << access_ptr;
+
+    auto buffer = LookupBufferFromData(Downcast<Var>(access_ptr_call->args[1]), func_);
+    CHECK(buffer) << "The buffer corresponding to data Var " << access_ptr_call->args[1]
+                  << " is not found";
+    int buffer_row_size = CheckAndGetBufferRowSize(buffer.value());
+
+    PrimExpr smem_offset = access_ptr_call->args[2] + (offset.defined() ? offset.value() : 0);
+
+    // Convert offset to 2-dimension, reindex it and convert it back
+    PrimExpr row_idx = floordiv(smem_offset, buffer_row_size);
+    PrimExpr col_idx = floormod(smem_offset, buffer_row_size);
+
+    auto new_indices = PermuteIndices(row_idx, col_idx, buffer_row_size);
+    auto new_offset = analyzer_->Simplify(new_indices[0] * buffer_row_size + new_indices[1]);
+
+    auto new_access_ptr = access_ptr_call.CopyOnWrite();
+    new_access_ptr->args.Set(2, new_offset);
+    return access_ptr_call;
+  }
+
+  PrimExpr VisitExpr_(const CallNode* op) final {
+    // Rewrite from/to shared or shared.dyn to/from local
+    auto call = Downcast<Call>(IRMutatorWithAnalyzer::VisitExpr_(op));
+
+    if (!permute_) {
+      return call;
+    }
+
+    if (!call->op.same_as(builtin::ptx_ldmatrix()) && !call->op.same_as(builtin::mma_store())) {
+      return call;
+    }
+
+    if (call->op.same_as(builtin::ptx_ldmatrix())) {
+      // form: T.ptx_ldmatrix(..., smem_ptr, smem_offset)
+      // smem_ptr: T.tvm_access_ptr(ptype, data, offset, extent, rw_mask)
+      auto access_ptr = call->args[5];
+      PrimExpr smem_offset = call->args[6];
+      auto new_access_ptr = HandleAccessPtrAndOffset(access_ptr, smem_offset);
```

Comment:
it would be better to record var to buffer mapping during visiting to avoid visit the whole prim func to find the buffer